### PR TITLE
Fix SugarCRM records not syncing - #7112 and #7138

### LINKED ISSUE #7112: Suite/Sugar integration doesn't update records (Mautic 2.15.00 /  SuiteCRM 7.10.10)
The connection plugin only creates new leads from Mautic to SuiteCRM, but fails to update them when some change in integration fields happens.

| Mautic 2.15.00 |
| SuiteCRM 7.10.10 |
| PHP 7.0 |
| Google Chrome |

Steps to reproduce
activate the plugin and choose:
Triggered action push contacts to integration
Pull contacts and/or companies from integration
Push contacts and/or companies to this integration
choose also to pool at least "leads"
map (at least) email address and first and last name, choosing Mautic to SuiteCRM as update direction for every field
create a lead in Mautic
give the following command through the console:
php -f console mautic:integration:synccontacts --integration="Sugarcrm" --time-interval="2 hours"
go to SuiteCRM and you'll see that the contact has been created
update the contact created in Mautic changing the first and last name
give again the command:
php -f console mautic:integration:synccontacts --integration="Sugarcrm" --time-interval="2 hours"

the console will give you the following output:

" Number of leads/contacts fetched: 1
Updating/creating leads from Mautic to Sugarcrm
Number of contacts pushed: 0 updated, 0 created, 0 had errors and 0 were ignored (likely duplicates or didn't match field criteria) "

So, the system wasn't able to update the record in SuiteCRM.

There is no error in the Mautic log file.

### LINKED ISSUE #7138: Sugar Mautic Sync only Syncing ONCE (not overwriting data)
## Bug Description
I have integrated the Mautic Sugar CRM Sync with Mautic. The sync is working Perfectly. However, there is one major flaw that no one seems to be talking about on the forums.... 

It will sync only ONCE and will not overwrite data after the first sync. 

A typical scenario --> Mautic gives 5 points and then pushes lead to Sugar --> Telemarketer calls this lead and then marks call status in CRM as Not available..... Not available then syncs from Sugar to Mautic. Ok all systems checked ---- except here is where the big flaw is --> next time the telemarketer calls that lead and updates Call Status to "Declined" --> this will not update from Sugar CRM to Mautic because it already pulled it last time as "Not Available" (It wont overwrite it" (And yes Sugar is the Master update direction)---> it is very important to overwrite data in the sync as data is constantly changing. 

In this example, we are wanting to create Segments in Mautic based on the Call Status in the CRM (which are constantly changing). 

| Q   | A
| --- | ---
| Mautic version | 2.15.0
| PHP version | PHP Version 7.0.32- 0ubuntu0.16.04.1 
| Browser |  CHROME 
| Sugar Version| Sugar Version 6.5.25 (Build 344) with SuiteCRM Version 7.10.9

### Steps to reproduce
1. Create a dropdown field in Sugar CRM and sync it with Mautic. Use the dropdown and select Not available from the dropdown list. Let the cron Run and update in Mautic
2. After "not available" has synced, go back into Sugar CRM , use the drop down on Call Status and select "declined" from the drop down list. 
3. Run the cron to update contacts. You will notice that in Mautic the Call Status is not updated to declined, it remains at the last sync (which is not available). 

 
### Log errors
With that being said. I do have some log errors in my log, but this I believe has nothing to do with the Sync not working because it is working flawlessly except for this issue. 

[2019-01-16 11:50:07] mautic.WARNING: PHP Warning - require(/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php): failed to open stream: No such file or directory - in file /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php - at line 209 [] []

[2019-01-16 11:50:07] mautic.WARNING: PHP Warning - require(/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php): failed to open stream: No such file or directory - in file /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php - at line 209 [] []

[2019-01-16 11:50:07] mautic.ERROR: PHP Notice: require(): Failed opening required '/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php' (include_path='.:/usr/share/php') - in file /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php - at line 209 [] []

[2019-01-16 11:50:07] mautic.ERROR: Symfony\Component\Debug\Exception\FatalErrorException: Notice: require(): Failed opening required '/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php' (include_path='.:/usr/share/php') - in file /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php - at line 209 [] []

[2019-01-16 11:50:07] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalErrorException: "Compile Error: require(): Failed opening required '/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php' (include_path='.:/usr/share/php')" at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php line 209 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalErrorException(code: 0): Compile Error: require(): Failed opening required '/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityPlugin.php' (include_path='.:/usr/share/php') at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php:209)"} []

[2019-01-16 01:35:06] mautic.NOTICE: RuntimeException: Unable to write in the "/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod" directory (uncaught exception) at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/symfony/framework-bundle/Command/CacheClearCommand.php line 67 while running console command `cache:clear` [] []

[2019-01-16 05:35:05] mautic.NOTICE: RuntimeException: Unable to write in the "/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod" directory (uncaught exception) at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/symfony/framework-bundle/Command/CacheClearCommand.php line 67 while running console command `cache:clear` [] []

[2019-01-16 10:35:05] mautic.NOTICE: RuntimeException: Unable to write in the "/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod" directory (uncaught exception) at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/symfony/framework-bundle/Command/CacheClearCommand.php line 67 while running console command `cache:clear` [] []

[2019-01-16 12:35:05] mautic.NOTICE: RuntimeException: Unable to write in the "/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod" directory (uncaught exception) at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/symfony/framework-bundle/Command/CacheClearCommand.php line 67 while running console command `cache:clear` [] []

[2019-01-16 14:35:05] mautic.NOTICE: RuntimeException: Unable to write in the "/var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/app/cache/prod" directory (uncaught exception) at /var/www/vhosts/<YOURURLHERE>/httpdocs/mautic/vendor/symfony/framework-bundle/Command/CacheClearCommand.php line 67 while running console command `cache:clear` [] []

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by tsummerer**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL | n/a
| Related developer documentation PR URL | n/a
| Issues addressed (#s or URLs) | #7112 #7138 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The current code prevents records from being being synced to CRM if the corresponding record has updates both in Mautic and CRM within the given Time Interval.

For example, if the interval is 15 minutes and the record was updated in both systems within the 15 minutes, then the changes will not be completely synced. 

This PR fixes that by delaying the save of the IntegrationEntity records until a more reasonable time.

Please see my comment in #7112 for a full explanation. 
tldr; The `getLeads` function sets the IntegrationEntity's 'last_sync_date' to Now. The `pushLeads` function only pushes Leads where the `date_modified` is GREATER than the `last_sync_date` - this is obviously never going to happen when the last_sync_date is only milliseconds ago. 
We need to wait to set the last_sync_date until later in the `pushLeads` process. 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Set up the Mautic/SugarCRM connector and map fields. Be sure that you have at least some fields going each direction. 
2. Create a Lead in Mautic
3. Run the `fetchleads` (or `synccontacts`) command : `mautic:integration:fetchleads --integration=Sugarcrm`
    This should create the Lead in CRM.
    You should have an Integration record under the Mautic Contact:
    Screen Shot

4. Modify the record in CRM. Change one of the 'sync to Mautic fields'
5. Modify the record in Mautic. Change one of the 'sync to CRM fields'
6. Run the `fetchleads` command again. 

Expected: The records are in sync. The CRM changes are reflected in Mautic and the Mautic changes in CRM. 

Actual : The records are not in sync. The CRM changes ARE reflected in Mautic, but the Mautic changes are NOT reflected in CRM. 

#### Steps to test this PR:
1. Load up this PR
2. Repeat steps above. You should receive the Expected results.

#### List deprecations along with the new alternative:
1. n/a

#### List backwards compatibility breaks:
1. n/a

### CODE REVIEW

**Comment by kuzmany** [comment]
@tsummerer Can you check code-style PHP 7.2 issues?

https://travis-ci.org/mautic/mautic/jobs/519369874#L814

File: plugins/MauticCrmBundle/Integration/SugarcrmIntegration.php [hunk 1]
```diff
@@ -1911,4 +1920,35 @@ public function convertSuiteCrmToMauticMultiSelect($suiteCrmMultiSelectStringToC
 
         return $convertedString;
     }
+
+
```
Comment (escopecz) [inline plugins/MauticCrmBundle/Integration/SugarcrmIntegration.php]: @tsummerer Travis CI is complaining about these 2 rows. There should be only one.

If you want the fixer to fix it for you then run `bin/php-cs-fixer fix -v`. But it may take a while.

**Comment by kuzmany** [comment]
Moved to next version.
Need more time for tests.

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 versions (except for major security purpose) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.0.1 candidate.**

## How to do?
1. Check if the bug still present in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please comment on your PR: “This PR only applies to Mautic 2.x”

**Please report results by commenting on your PR to make us administration easier.**

_In case your bugfix only apply to Mautic 2, we'll consider adding it in an extra Mautic 2 version._
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by npracht** [comment]
Hello @tsummerer could you please test mautic 3.0.0-beta2 and tell me if the issue is still existing?
- If no, i'll tag this issue at `M2 only relevant`
- If yes, could you please rebase your branch against 3.x branch ?

Waiting for your feedback, i would very love to merge it in 3.0.1 if still relevant. Thanks !

**Comment by RCheesley** [comment]
@tsummerer could you take a look at the conflict please? Thanks!

**Comment by tsummerer** [comment]
I'm sorry folks, but I am no longer using the Mautic/Sugar plugin. There were a variety of other hard-to-reproduce-but-persistent issues and we ended up developing a different solution (unfortunately it relied heavily on our specific architecture and wouldn't have been well suited to open-source).
I won't be able to set up and run a decent test for quite some time. If there's anything I can do to allow someone else to carry it home, I'd be happy to. Sorry.
